http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/conn/IdleConnectionReaper.java

```diff
@@ -15,181 +15,163 @@
+
+    private volatile ReaperTask reaperTask;
 
-    /**
-     * Private constructor - singleton pattern.
-     */
     private IdleConnectionReaper() {
-        super("java-sdk-http-connection-reaper");
-        setDaemon(true);
+        this.connectionManagers =  new ConcurrentHashMap<>();
+
+        this.executorServiceSupplier = () -> {
+            ExecutorService e = Executors.newSingleThreadExecutor(r -> {
+                Thread t = new Thread(r, "idle-connection-reaper");
+                t.setDaemon(true);
+                return t;
+            });
+            return e;
+        };
+
+        this.sleepPeriod = Duration.ofMinutes(1).toMillis();
     }
 
-    /**
-     * Registers the given connection manager with this reaper.
-     *
-     * @return true if the connection manager has been successfully registered; false otherwise.
-     * @deprecated By {@link #registerConnectionManager(HttpClientConnectionManager, long)}.
-     */
-    @Deprecated
-    public static boolean registerConnectionManager(HttpClientConnectionManager connectionManager) {
-        return registerConnectionManager(connectionManager, DEFAULT_MAX_IDLE_MILLIS);
+    @SdkTestInternalApi
+    IdleConnectionReaper(Map<HttpClientConnectionManager, Long> connectionManagers,
+                         Supplier<ExecutorService> executorServiceSupplier,
+                         long sleepPeriod) {
+
+        this.connectionManagers = connectionManagers;
+        this.executorServiceSupplier = executorServiceSupplier;
+        this.sleepPeriod = sleepPeriod;
     }
 
     /**
-     * Registers the given connection manager with this reaper;
+     * Register the connection manager with this reaper.
      *
-     * @param connectionManager Connection manager to register
-     * @param maxIdleInMs       Max idle connection timeout in milliseconds for this connection manager.
-     * @return true if the connection manager has been successfully registered; false otherwise.
+     * @param manager The connection manager.
+     * @param maxIdleTime The maximum time connections in the connection manager are to remain idle before being reaped.
+     * @return {@code true} If the connection manager was not previously registered with this reaper, {@code false}
+     * otherwise.
      */
-    public static boolean registerConnectionManager(HttpClientConnectionManager connectionManager, long maxIdleInMs) {
-        if (instance == null) {
-            synchronized (IdleConnectionReaper.class) {
-                if (instance == null) {
-                    IdleConnectionReaper newInstance = new IdleConnectionReaper();
-                    newInstance.start();
-                    instance = newInstance;
-                }
-            }
+    public boolean registerConnectionManager(HttpClientConnectionManager manager, long maxIdleTime) {
+        boolean notPreviouslyRegistered = connectionManagers.put(manager, maxIdleTime) == null;
+        if (notPreviouslyRegistered) {
```

Comment:
do we need the check of not previously registered since we are checking if the exec is null?

Reply:
Nope, don't think it's necessary.